### PR TITLE
pipeline refactor and cluster watch

### PR DESCRIPTION
**Pull request by chengjoey**
#### What type of this PR

Add one of the following kinds:
/kind feature

#### What this PR does / why we need it:
pipeline refactor, add three executor in scheduler plugin(k8sjob, k8sflink, k8sspark), dcos and edas will dispath to scheduler module
get cluster from cluster-manager module, and register cluster hook
#### Specified Reviewers:

/assign @your-reviewer

### CODE REVIEW

File: apistructs/cluster.go [hunk 1]
```diff
@@ -36,6 +36,18 @@ const (
...
 )
 
+const (
+	ClusterActionCreate = "create"
+	ClusterActionUpdate = "update"
+	ClusterActionDelete = "delete"
+)
+
+const (
+	ClusterTypeDcos = "dcos"
```
Comment (sfwn) [inline apistructs/cluster.go:46]: Can you reuse definitions at line 20~22 in this file.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/initialize.go [hunk 2]
```diff
@@ -221,6 +232,23 @@ func registerSnippetClient(dbclient *dbclient.Client) {
 	})
 }
 
+func registerClusterHook(bdl *bundle.Bundle) {
+	ev := apistructs.CreateHookRequest{
+		Name:   "pipeline_watch_cluster_changed",
+		Events: []string{bundle.ClusterEvent},
+		URL:    strutil.Concat("http://", discover.Pipeline(), "/clusterhook"),
```
Comment (sfwn) [inline modules/pipeline/initialize.go:239]: Use same const variable with endpoints.

File: modules/pipeline/endpoints/clusterinfo.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	"github.com/sirupsen/logrus"
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/pkg/http/httpserver"
+)
+
+func (e *Endpoints) clusterHook(ctx context.Context, r *http.Request, vars map[string]string) (httpserver.Responser, error) {
+	req := apistructs.ClusterEvent{}
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		errStr := fmt.Sprintf("decode clusterhook request fail: %v", err)
```
Comment (sfwn) [inline modules/pipeline/endpoints/clusterinfo.go:31]: failed to xxx

File: modules/pipeline/initialize.go (hunk 2 above)
Comment (sfwn) [reply modules/pipeline/initialize.go:239]: how about multi instance?

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/manager.go [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+		name = types.Name(fmt.Sprintf("%sfor%s", cluster.Name, k8sspark.Kind))
+		if _, exist := m.executors[name]; exist {
+			delete(m.executors, name)
+		}
+	default:
+
+	}
+}
+
+func (m *Manager) updateExecutor(cluster apistructs.ClusterInfo) error {
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/manager.go:206]: duplicated with addExecutor.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: go.sum [hunk 5]
```diff
@@ -228,6 +228,8 @@ github.com/cheekybits/is v0.0.0-20150225183255-68e9c0620927/go.mod h1:h/aW8ynjgk
 github.com/cheggaaa/pb/v3 v3.0.1/go.mod h1:SqqeMF/pMOIu3xgGoxtPYhMNQP258xE4x/XRTYua+KU=
 github.com/cheggaaa/pb/v3 v3.0.4 h1:QZEPYOj2ix6d5oEg63fbHmpolrnNiwjUsk+h74Yt4bM=
 github.com/cheggaaa/pb/v3 v3.0.4/go.mod h1:7rgWxLrAUcFMkvJuv09+DYi7mMUYi8nO9iOWcvGJPfw=
+github.com/chengjoey/flink-on-k8s-operator v0.0.1 h1:a4zrgATUtbQXFrfz0OAH1C9crs4fojzp2k25oNlgfl0=
```
Comment (sfwn) [inline go.sum:231]: https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/issues/460

File: modules/pipeline/initialize.go [hunk 6]
```diff
@@ -194,6 +201,9 @@ func do() (*httpserver.Server, error) {
 		return nil, err
 	}
 
+	// register cluster hook
+	registerClusterHook(bdl)
```
Comment (sfwn) [inline modules/pipeline/initialize.go:205]: Put registerClusterHook into pkg clusterinfo is better?
Put these two methods together at least.

Comment (sfwn) [reply modules/pipeline/initialize.go:205]: Or only one method here such as `doClusterAbout`.

File: modules/pipeline/initialize.go [hunk 7]
```diff
@@ -219,6 +229,23 @@ func registerSnippetClient(dbclient *dbclient.Client) error {
...
+		URL:    strutil.Concat("http://", conf.PodIP(), endpoints.ClusterHookApiPath),
+		Active: true,
+		HookLocation: apistructs.HookLocation{
+			Org:         "-1",
+			Project:     "-1",
+			Application: "-1",
+		},
+	}
+	if err := bdl.CreateWebhook(ev); err != nil {
+		logrus.Warnf("failed to register watch cluster changed event, %v", err)
```
Comment (sfwn) [inline modules/pipeline/initialize.go:245]: Return err and handle err at invoker side.

Comment (sfwn) [reply modules/pipeline/initialize.go:245]: And I think pipeline should panic if registerClusterHook failed at bootsstrap.

Comment (sfwn) [reply modules/pipeline/initialize.go:245]: Each time  the pod ip is different.
How about old pod ips already registered? Maybe eventbox will continue send msgs to them but failed of course. And the subscribers(pod ips) will be more and more ?

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/initialize.go (hunk 7 above)
Comment (sfwn) [reply modules/pipeline/initialize.go:245]: If so, maybe we should continue register with service ip and all instances listen from etcd.

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/manager.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+		}
+	}
+
+	eventChan, err := clusterinfo.RegisterClusterEvent()
+	if err != nil {
+		return err
+	}
+	go m.listenClusterEventSync(context.Background(), eventChan)
+
+	logrus.Info("pipengine task executor manager Initialize Done .")
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/manager.go:67]: pipeline scheduler task executor

Comment (sfwn) [reply modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/manager.go:67]: Check all others.

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sflink/flink.go [hunk 9]
```diff
@@ -0,0 +1,259 @@
...
+
+	"github.com/erda-project/erda/apistructs"
+	"github.com/erda-project/erda/modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/types"
+	"github.com/erda-project/erda/modules/pipeline/pipengine/actionexecutor/plugins/scheduler/logic"
+	"github.com/erda-project/erda/modules/pipeline/spec"
+	"github.com/erda-project/erda/pkg/strutil"
+)
+
+const (
+	DiceRootDomainKEY  = "DICE_ROOT_DOMAIN"
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sflink/flink.go:36]: Key

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sjob/error.go [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+// FITNESS FOR A PARTICULAR PURPOSE.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+package k8sjob
+
+var (
+	errPullImage               = "拉取镜像失败"
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/plugins/k8sjob/error.go:17]: Later with i18n

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/types/define.go [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+package types
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gogap/errors"
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/executor/types/define.go:20]: Check all error pkg, use fmt.Errorf or unified github.com/pkg/errors.

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/logic/shcedule.go [hunk 12]
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2021 Terminus, Inc.
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/logic/shcedule.go:1]: Typo of file name schedule.

File: modules/pipeline/pipengine/actionexecutor/plugins/scheduler/scheduler.go [hunk 13]
```diff
@@ -23,14 +23,17 @@ import (
 	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 
+	"github.com/erda-project/erda/modules/pipeline/pipengine/actionexecutor/plugins/scheduler/logic"
+
```
Comment (sfwn) [inline modules/pipeline/pipengine/actionexecutor/plugins/scheduler/scheduler.go:27]: no need new line.

File: modules/pipeline/pipengine/reconciler/taskrun/taskop/start.go [hunk 14]
```diff
@@ -159,6 +164,19 @@ func (s *start) TuneTriggers() taskrun.TaskOpTuneTriggers {
 	}
 }
 
+// injectJobID save flink, spark job id after start
+func injectJobID(tr *taskrun.TaskRun, data interface{}) error {
+	if data == nil {
+		return nil
+	}
+	jobID, ok := data.(string)
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/taskrun/taskop/start.go:172]: You should check task executor type here.
Not all string type data is job id.

File: modules/pipeline/pipengine/reconciler/taskrun/taskop/start.go [hunk 15]
```diff
@@ -159,6 +164,19 @@ func (s *start) TuneTriggers() taskrun.TaskOpTuneTriggers {
 	}
 }
 
+// injectJobID save flink, spark job id after start
+func injectJobID(tr *taskrun.TaskRun, data interface{}) error {
```
Comment (sfwn) [inline modules/pipeline/pipengine/reconciler/taskrun/taskop/start.go:168]: injectJobIDForFlinkSpark is better.

File: modules/pipeline/services/pipelinesvc/clusterinfo.go [hunk 16]
```diff
@@ -18,9 +18,11 @@ import (
 	"strings"
 	"time"
 
+	"github.com/gogap/errors"
```
Comment (sfwn) [inline modules/pipeline/services/pipelinesvc/clusterinfo.go:21]: the same.

File: modules/pipeline/services/pipelinesvc/clusterinfo.go [hunk 17]
```diff
@@ -57,3 +59,19 @@ func (s *PipelineSvc) retryQueryClusterInfo(clusterName string, pipelineID uint6
 
 	return result, queryErr
 }
+
+// ClusterHook listen and dispatch cluster event from eventbox
+func (s *PipelineSvc) ClusterHook(clusterEvent apistructs.ClusterEvent) error {
+	if !strutil.Equal(clusterEvent.Content.Type, apistructs.K8S, true) &&
```
Comment (sfwn) [inline modules/pipeline/services/pipelinesvc/clusterinfo.go:65]: Use switch grammar is more clear here.

File: modules/pipeline/spec/pipeline_task.go [hunk 18]
```diff
@@ -95,6 +95,8 @@ type PipelineTaskExtra struct {
 	LoopOptions *apistructs.PipelineTaskLoopOptions `json:"loopOptions,omitempty"` // 开始执行后保证不为空
 
 	AppliedResources apistructs.PipelineAppliedResources `json:"appliedResources,omitempty"`
+
+	JobID string `json:"jobID,omitempty"` // flink, spark job id
```
Comment (sfwn) [inline modules/pipeline/spec/pipeline_task.go:99]: Concrete field name.

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/pipengine/reconciler/taskrun/taskop/start.go (hunk 14 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/reconciler/taskrun/taskop/start.go:172]: only dcos return job id, so remove inJectJobID

File: modules/pipeline/initialize.go (hunk 6 above)
Comment (sfwn) [reply modules/pipeline/initialize.go:202]: Move `registerClusterHook` into `clusterinfo.Initialize`

**Review by sfwn — CHANGES_REQUESTED** [review]

File: modules/pipeline/pipengine/reconciler/taskrun/taskop/start.go (hunk 15 above)
Comment (chengjoey) [reply modules/pipeline/pipengine/reconciler/taskrun/taskop/start.go]: only dcos return str job id,already remove injectJobID

**Review by sfwn — APPROVED** [review]

**Pull request merged**